File: src/lib/llmProxy.ts
/**
 * OpenAI 兼容 Chat Completions 同源代理。
 * 浏览器把用户自己的 Key 放在 Authorization；本函数只转发，不落盘。
 */
export type LlmChatRole = "system" | "user" | "assistant" | "tool";

export type LlmToolCallPayload = {
  id: string;
  type?: "function";
  function: {
    name: string;
    arguments: string;
  };
};

export type LlmChatMessage = {
  role: LlmChatRole;
  content: string;
  name?: string;
  tool_call_id?: string;
  tool_calls?: LlmToolCallPayload[];
};

export type LlmToolDefinition = {
  type: "function";
  function: {
    name: string;
    description: string;
    parameters: Record<string, unknown>;
  };
};

export type LlmChatProxyBody = {
  baseUrl: string;
  model: string;
  messages: LlmChatMessage[];
  temperature?: number;
  jsonMode?: boolean;
  maxTokens?: number;
  tools?: LlmToolDefinition[];
  toolChoice?: "auto" | "none";
};

const LOCAL_HOSTS = new Set(["localhost", "127.0.0.1", "[::1]", "::1"]);
const MAX_MESSAGES = 48;
const MAX_MESSAGE_CHARS = 20_000;
/** 表格导入等长输出；供应商若更低会自行截断或报错。 */
export const MAX_COMPLETION_TOKENS = 16_384;
const REQUEST_TIMEOUT_MS = 120_000;

export const normalizeChatCompletionsUrl = (baseUrl: string) => {
  const trimmed = baseUrl.trim().replace(/\/+$/, "");
  if (/\/chat\/completions$/i.test(trimmed)) return trimmed;
  if (/\/chatcompletion_v2$/i.test(trimmed)) return trimmed;
  try {
    const { pathname } = new URL(trimmed);
    const path = pathname.replace(/\/+$/, "") || "/";
    // 已是完整 endpoint（如 MiniMax /v1/text/chatcompletion_v2），不再拼 /chat/completions
    if (path !== "/" && path !== "/v1" && !path.endsWith("/v1")) return trimmed;
  } catch {
    /* fall through */
  }
  return `${trimmed}/chat/completions`;
};

export const isAllowedLlmBaseUrl = (baseUrl: string) => {
  try {
    const url = new URL(baseUrl.trim());
    if (url.username || url.password) return false;
    if (url.protocol === "https:") return Boolean(url.hostname);
    if (url.protocol === "http:") return LOCAL_HOSTS.has(url.hostname);
    return false;
  } catch {
    return false;
  }
};

const jsonError = (message: string, status: number) =>
  new Response(JSON.stringify({ error: message }), {
    status,
    headers: {
      "Content-Type": "application/json; charset=utf-8",
      "Cache-Control": "no-store",
    },
  });

const isChatRole = (value: unknown): value is LlmChatRole =>
  value === "system" || value === "user" || value === "assistant" || value === "tool";

const normalizeMessageContent = (value: unknown) => {
  if (typeof value === "string") return value;
  if (value == null) return "";
  return null;
};

const sanitizeToolCalls = (value: unknown): LlmToolCallPayload[] | undefined => {
  if (!Array.isArray(value) || !value.length) return undefined;
  const calls: LlmToolCallPayload[] = [];
  for (const item of value) {
    if (!item || typeof item !== "object") continue;
    const record = item as {
      id?: unknown;
      type?: unknown;
      function?: { name?: unknown; arguments?: unknown };
    };
    const name = record.function?.name;
    if (typeof name !== "string" || !name.trim()) continue;
    const args = record.function?.arguments;
    calls.push({
      id:
        typeof record.id === "string" && record.id.trim() ? record.id : `call_${calls.length + 1}`,
      type: "function",
      function: {
        name: name.trim(),
        arguments: typeof args === "string" ? args : JSON.stringify(args ?? {}),
      },
    });
  }
  return calls.length ? calls : undefined;
};

export const handleLlmChatRequest = async (request: Request): Promise<Response> => {
  const apiKey =
    request.headers
      .get("authorization")
      ?.replace(/^Bearer\s+/i, "")
      .trim() ?? "";
  if (!apiKey) return jsonError("缺少 API Key。", 401);

  let body: LlmChatProxyBody;
  try {
    body = (await request.json()) as LlmChatProxyBody;
  } catch {
    return jsonError("请求体不是合法 JSON。", 400);
  }

  if (!body || typeof body !== "object") return jsonError("请求体无效。", 400);
  if (typeof body.baseUrl !== "string" || !body.baseUrl.trim()) {
    return jsonError("请提供 API Base URL。", 400);
  }
  if (!isAllowedLlmBaseUrl(body.baseUrl)) {
    return jsonError("API 地址只允许 https，或本机 http://localhost。", 400);
  }
  if (typeof body.model !== "string" || !body.model.trim()) {
    return jsonError("请提供模型名。", 400);
  }
  if (!Array.isArray(body.messages) || body.messages.length === 0) {
    return jsonError("messages 不能为空。", 400);
  }
  if (body.messages.length > MAX_MESSAGES) return jsonError("messages 过多。", 400);

  for (const message of body.messages) {
    const content = normalizeMessageContent(message?.content);
    if (!message || !isChatRole(message.role) || content == null) {
      return jsonError("message.role / content 无效。", 400);
    }
    if (content.length > MAX_MESSAGE_CHARS) {
      return jsonError("单条 message 过长。", 400);
    }
    message.content = content;
  }

  const payload: Record<string, unknown> = {
    model: body.model.trim(),
    messages: body.messages.map((item) => {
      const next: Record<string, unknown> = { role: item.role, content: item.content };
      if (item.name) next.name = item.name;
      if (item.tool_call_id) next.tool_call_id = item.tool_call_id;
      if (item.tool_calls?.length) next.tool_calls = sanitizeToolCalls(item.tool_calls);
      return next;
    }),
    temperature: typeof body.temperature === "number" ? body.temperature : 0,
    stream: false,
  };
  const tools = Array.isArray(body.tools) ? body.tools : [];
  const hasTools = tools.length > 0;
  if (hasTools) {
    payload.tools = tools.slice(0, 16);
    if (body.toolChoice === "none" || body.toolChoice === "auto") {
      payload.tool_choice = body.toolChoice;
    }
  }
  if (body.jsonMode === true || (body.jsonMode !== false && !hasTools)) {
    payload.response_format = { type: "json_object" };
  }
  if (typeof body.maxTokens === "number" && body.maxTokens > 0) {
    payload.max_tokens = Math.min(Math.floor(body.maxTokens), MAX_COMPLETION_TOKENS);
  }

  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), REQUEST_TIMEOUT_MS);

  try {
    const upstream = await fetch(normalizeChatCompletionsUrl(body.baseUrl), {
      method: "POST",
      headers: {
        Authorization: `Bearer ${apiKey}`,
        "Content-Type": "application/json",
        Accept: "application/json",
      },
      body: JSON.stringify(payload),
      signal: controller.signal,
    });
    const text = await upstream.text();
    return new Response(text, {
      status: upstream.status,
      headers: {
        "Content-Type": upstream.headers.get("Content-Type") ?? "application/json; charset=utf-8",
        "Cache-Control": "no-store",
      },
    });
  } catch (error) {
    const aborted = error instanceof Error && error.name === "AbortError";
    return jsonError(aborted ? "模型请求超时。" : "无法连接模型接口。", aborted ? 504 : 502);
  } finally {
    clearTimeout(timer);
  }
};

File: desktop/internal/llmproxy/llmproxy.go
// Package llmproxy ports src/lib/llmProxy.ts for the Wails asset server.
// Keep URL allowlist, timeouts, and payload limits in sync with the TypeScript copy.
package llmproxy

import (
	"bytes"
	"encoding/json"
	"io"
	"net/http"
	"net/url"
	"strings"
	"time"
)

const (
	maxMessages         = 24
	maxMessageChars     = 20_000
	maxCompletionTokens = 16_384
	requestTimeout      = 120 * time.Second
)

var localHosts = map[string]struct{}{
	"localhost": {},
	"127.0.0.1": {},
	"[::1]":     {},
	"::1":       {},
}

type chatMessage struct {
	Role    string `json:"role"`
	Content string `json:"content"`
}

type chatBody struct {
	BaseURL     string        `json:"baseUrl"`
	Model       string        `json:"model"`
	Messages    []chatMessage `json:"messages"`
	Temperature *float64      `json:"temperature"`
	JSONMode    *bool         `json:"jsonMode"`
	MaxTokens   *int          `json:"maxTokens"`
}

func NormalizeChatCompletionsURL(baseURL string) string {
	trimmed := strings.TrimRight(strings.TrimSpace(baseURL), "/")
	lower := strings.ToLower(trimmed)
	if strings.HasSuffix(lower, "/chat/completions") {
		return trimmed
	}
	if strings.HasSuffix(lower, "/chatcompletion_v2") {
		return trimmed
	}
	parsed, err := url.Parse(trimmed)
	if err == nil {
		path := strings.TrimRight(parsed.Path, "/")
		if path == "" {
			path = "/"
		}
		if path != "/" && path != "/v1" && !strings.HasSuffix(path, "/v1") {
			return trimmed
		}
	}
	return trimmed + "/chat/completions"
}

func IsAllowedLlmBaseURL(baseURL string) bool {
	parsed, err := url.Parse(strings.TrimSpace(baseURL))
	if err != nil || parsed.Host == "" {
		return false
	}
	if parsed.User != nil {
		return false
	}
	switch parsed.Scheme {
	case "https":
		return parsed.Hostname() != ""
	case "http":
		host := parsed.Hostname()
		_, ok := localHosts[host]
		return ok
	default:
		return false
	}
}

func jsonError(w http.ResponseWriter, message string, status int) {
	w.Header().Set("Content-Type", "application/json; charset=utf-8")
	w.Header().Set("Cache-Control", "no-store")
	w.WriteHeader(status)
	_ = json.NewEncoder(w).Encode(map[string]string{"error": message})
}

func isChatRole(role string) bool {
	return role == "system" || role == "user" || role == "assistant"
}

func HandleLlmChat(w http.ResponseWriter, r *http.Request) {
	if r.Method == http.MethodOptions {
		w.Header().Set("Cache-Control", "no-store")
		w.WriteHeader(http.StatusNoContent)
		return
	}
	if r.Method != http.MethodPost {
		jsonError(w, "Method not allowed", http.StatusMethodNotAllowed)
		return
	}

	apiKey := strings.TrimSpace(r.Header.Get("Authorization"))
	if len(apiKey) >= 7 && strings.EqualFold(apiKey[:7], "Bearer ") {
		apiKey = strings.TrimSpace(apiKey[7:])
	}
	if apiKey == "" {
		jsonError(w, "缺少 API Key。", http.StatusUnauthorized)
		return
	}

	var body chatBody
	decoder := json.NewDecoder(r.Body)
	if err := decoder.Decode(&body); err != nil {
		jsonError(w, "请求体不是合法 JSON。", http.StatusBadRequest)
		return
	}

	if strings.TrimSpace(body.BaseURL) == "" {
		jsonError(w, "请提供 API Base URL。", http.StatusBadRequest)
		return
	}
	if !IsAllowedLlmBaseURL(body.BaseURL) {
		jsonError(w, "API 地址只允许 https，或本机 http://localhost。", http.StatusBadRequest)
		return
	}
	if strings.TrimSpace(body.Model) == "" {
		jsonError(w, "请提供模型名。", http.StatusBadRequest)
		return
	}
	if len(body.Messages) == 0 {
		jsonError(w, "messages 不能为空。", http.StatusBadRequest)
		return
	}
	if len(body.Messages) > maxMessages {
		jsonError(w, "messages 过多。", http.StatusBadRequest)
		return
	}
	for _, message := range body.Messages {
		if !isChatRole(message.Role) {
			jsonError(w, "message.role / content 无效。", http.StatusBadRequest)
			return
		}
		if len(message.Content) > maxMessageChars {
			jsonError(w, "单条 message 过长。", http.StatusBadRequest)
			return
		}
	}

	temperature := 0.0
	if body.Temperature != nil {
		temperature = *body.Temperature
	}
	payload := map[string]any{
		"model":       strings.TrimSpace(body.Model),
		"messages":    body.Messages,
		"temperature": temperature,
		"stream":      false,
	}
	jsonMode := true
	if body.JSONMode != nil {
		jsonMode = *body.JSONMode
	}
	if jsonMode {
		payload["response_format"] = map[string]string{"type": "json_object"}
	}
	if body.MaxTokens != nil && *body.MaxTokens > 0 {
		tokens := *body.MaxTokens
		if tokens > maxCompletionTokens {
			tokens = maxCompletionTokens
		}
		payload["max_tokens"] = tokens
	}

	raw, err := json.Marshal(payload)
	if err != nil {
		jsonError(w, "无法连接模型接口。", http.StatusBadGateway)
		return
	}

	req, err := http.NewRequestWithContext(r.Context(), http.MethodPost, NormalizeChatCompletionsURL(body.BaseURL), bytes.NewReader(raw))
	if err != nil {
		jsonError(w, "无法连接模型接口。", http.StatusBadGateway)
		return
	}
	req.Header.Set("Authorization", "Bearer "+apiKey)
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Accept", "application/json")

	client := &http.Client{Timeout: requestTimeout}
	resp, err := client.Do(req)
	if err != nil {
		if r.Context().Err() != nil || strings.Contains(strings.ToLower(err.Error()), "timeout") {
			jsonError(w, "模型请求超时。", http.StatusGatewayTimeout)
			return
		}
		jsonError(w, "无法连接模型接口。", http.StatusBadGateway)
		return
	}
	defer resp.Body.Close()
	text, err := io.ReadAll(resp.Body)
	if err != nil {
		jsonError(w, "无法连接模型接口。", http.StatusBadGateway)
		return
	}
	contentType := resp.Header.Get("Content-Type")
	if contentType == "" {
		contentType = "application/json; charset=utf-8"
	}
	w.Header().Set("Content-Type", contentType)
	w.Header().Set("Cache-Control", "no-store")
	w.WriteHeader(resp.StatusCode)
	_, _ = w.Write(text)
}

func Middleware(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		path := r.URL.Path
		if q := strings.Index(path, "?"); q >= 0 {
			path = path[:q]
		}
		if path != "/api/llm/chat" {
			next.ServeHTTP(w, r)
			return
		}
		HandleLlmChat(w, r)
	})
}
